### PR TITLE
fix: Fix quick cast keys

### PR DESCRIPTION
**Pull request by ryanzhoudev**
- Copies artemis implementation
- Redid the checks that happen before starting a spell
    - No more mana check as the mana table is also gone
- Disables totem highlights for now as it sometimes hits non-totem objects

### CODE REVIEW

**Comment by byBackfish** [comment]
Sounds good. Used it for an hour or so and didn't run into problems. Removing the totem glow for now is also nice

File: src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java [hunk 1]
```diff
@@ -41,221 +33,129 @@ public class QuickCastManager {
...
-    private static boolean checkKey(Packet<?> input, int pos, boolean clickType, boolean isLowLevel) {
-        boolean[] spell;
+    public static void castSecondSpell() {
+        if (checkSpellCastRequest(2) != CastCheckStatus.OK) return;
 
-        SpellData data = PlayerInfo.get(SpellData.class);
-        if (isLowLevel) {
-            SPacketTitle title = (SPacketTitle) input;
-            if (title.getType() != SPacketTitle.Type.SUBTITLE) return false;
+        if (PlayerInfo.get(CharacterData.class).getCurrentClass() == ClassType.ARCHER) {
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java]: Can't queueSpell invert the clicks if current class is archer?

Comment (ryanzhoudev) [reply src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java]: Done

**Comment by kristofbolyai** [comment]
@donkeyblaster copy artemis' check if spell casting is in progress?

**Comment by bashbers** [comment]
I would love to see this PR implemented. The shortcuts not working rn is kinda annoying.

**Comment by byBackfish** [comment]
> I would love to see this PR implemented. The shortcuts not working rn is kinda annoying.

It will be implemented soon, but it might take another day or two (hopefully)

File: src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java [hunk 2]
```diff
@@ -26,236 +29,218 @@
...
 
-public class QuickCastManager {
+public class QuickCastManager implements Listener {
 
     private static final CPacketAnimation leftClick = new CPacketAnimation(EnumHand.MAIN_HAND);
     private static final CPacketPlayerTryUseItem rightClick = new CPacketPlayerTryUseItem(EnumHand.MAIN_HAND);
-    private static final CPacketPlayerDigging releaseClick = new CPacketPlayerDigging(CPacketPlayerDigging.Action.RELEASE_USE_ITEM, BlockPos.ORIGIN, EnumFacing.DOWN);
-
-    private static final int[] spellUnlock = { 1, 6, 10, 10 };
+    // private static final CPacketPlayerDigging releaseClick = new CPacketPlayerDigging(CPacketPlayerDigging.Action.RELEASE_USE_ITEM, BlockPos.ORIGIN, EnumFacing.DOWN);
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java]: Why keep this

File: src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java [hunk 3]
```diff
@@ -26,236 +29,218 @@
...
+
+    private enum CastCheckStatus {
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java]: Please move the enum to the end of the file

File: src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java [hunk 4]
```diff
@@ -26,236 +29,218 @@
...
+
+        NetHandlerPlayClient connection = McIf.mc().getConnection();
+        if (connection == null) return;
+
+        if (slotChanged) connection.sendPacket(new CPacketHeldItemChange(lastSelectedSlot));
+        connection.sendPacket(spellPacketQueue.remove());
+        if (slotChanged) connection.sendPacket(new CPacketHeldItemChange(currentSelectedSlot));
+
+        // Number of ticks to delay between spell cast button inputs
+        if (!spellPacketQueue.isEmpty()) packetQueueCountdown = 4;
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java]: The 4 could be extracted into a named const for readability.

File: src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java [hunk 5]
```diff
@@ -26,236 +29,218 @@
...
     }
 
-    public static void castFourthSpell() {
-        if (PlayerInfo.get(CharacterData.class).getCurrentClass() == ClassType.ARCHER) {
-            queueSpell(4, SPELL_LEFT, SPELL_LEFT, SPELL_RIGHT);
-            return;
+    private static void tryUpdateSpell(String text) {
+        List<Boolean> spell = getSpellFromString(text);
+        if (spell == null) return;
+        if (spellInProgress.equals(spell)) return;
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java:148]: This is a weird check. I am not sure if this blocks fast recasting...

We can keep it though.

File: src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java [hunk 6]
```diff
@@ -26,236 +29,218 @@
...
-        //Is the current combat level enough to use the weapon
-        boolean combatLvlMinReached = false;
-        //If there is a spell point requirement that is not reached, store it and print it later
-        String notReachedSpellPointRequirements = null;
+        return spell;
+    }
 
-        if (!isWeapon)
-        {
+    private static CastCheckStatus checkSpellCastRequest() {
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java:176]: Kinda weird thing that a method for checking state is also printing the errors. 

If you want, we could refactor this, and print errors where we call this. Something like `if (CastCheckStatus != Ok) checkStatus.printError()`

Comment (kristofbolyai) [reply src/main/java/com/wynntils/modules/utilities/managers/QuickCastManager.java:176]: Not an approval blocking request though.

File: src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java [hunk 7]
```diff
@@ -855,11 +855,11 @@ public void onTotemEvent(SpellEvent.TotemActivated e) {
...
-        List<EntityArmorStand> possibleTotems = McIf.mc().world.getEntitiesWithinAABB(EntityArmorStand.class,
-                new AxisAlignedBB(e.getLocation().getX() - 1, e.getLocation().getY(), e.getLocation().getZ() - 1, e.getLocation().getX() + 1, e.getLocation().getY() + 3, e.getLocation().getZ() + 1));
-        possibleTotems.forEach(entityArmorStand -> {
-            if (entityArmorStand.getDisplayName().getUnformattedText().equals("Armor Stand")) entityArmorStand.setGlowing(true);
-        });
+//        List<EntityArmorStand> possibleTotems = McIf.mc().world.getEntitiesWithinAABB(EntityArmorStand.class,
+//                new AxisAlignedBB(e.getLocation().getX() - 1, e.getLocation().getY(), e.getLocation().getZ() - 1, e.getLocation().getX() + 1, e.getLocation().getY() + 3, e.getLocation().getZ() + 1));
+//        possibleTotems.forEach(entityArmorStand -> {
+//            if (entityArmorStand.getDisplayName().getUnformattedText().equals("Armor Stand")) entityArmorStand.setGlowing(true);
+//        });
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/overlays/OverlayEvents.java]: I don't know why to keep commented code. VCS can easily show us the old code.

**Review by kristofbolyai — APPROVED** [review]

**Review by Incompleteusern — APPROVED** [review]

**Pull request merged**